### ISSUE TITLE
Try to convert complete codebase to slog-rs?

### ISSUE THREAD

**Issue by matthiasbeyer** [matthiasbeyer/imag#671]
I would like if someone would try to convert the whole codebase to use [slog-rs](https://github.com/dpc/slog-rs) rather than [log](https://crates.io/crates/log) and check how it works out.

Disclaimer: If I do not like slog, I won't merge. It would be a nice experiement anyways and maybe it is better than `log`.

Lets check it out!

**Comment by neithernut** [comment]
It looks like they'll do a 1.0 release soon. A PR switching to slog-rs should use that release.

**Comment by matthiasbeyer** [comment]
@neithernut of course! That's also why I'm posting this issue. I've discovered [this reddit](https://www.reddit.com/r/rust/comments/4zncbu/slog_trivial_to_convert_to_with_slogenvlogger/) (I knew the crate before) and because of the approaching release, I posted this very issue... might be worth checking out, so we get better structure in our logging stuff. What do you think, @neithernut?

**Comment by neithernut** [comment]
Logging libraries of any kind tend to overkill the problem. (e.g. logging to extra locations when it's really the job of the syslog deamon to dispatch logs). I'm also sceptic of using async IO for logging. Hierarchical logging is a nice feature to have though, /if/ it is done right (inb4 lost context because it is only enabled on trace/debug or spamming the log with context information that is not needed).

**Comment by dpc** [comment]
Hi. 

I'd be happy to help on figuring out way to make using `slog` in libraries work well. Things like fallback to `log`, opting in/out of slog using cargo features, etc.

Saying that,  I'd say contextual information of logging statements is a big point of structered logging. By manually dealing with `Logger` object, context might correspond to the loggical flow, instead of being reduced to module,source file and line. If `imag` library itself does not have useful contextual information and complex flow (eg. own threads, long running job & instances), then there might be not much value added by switching to structured logging. `slog` users themselves can use `slog-stdlog ` and scopes to redirect `log` standard logging statements to `slog`.

@neithernut:

I'd say applications usually know better what do with their logging, while syslog is more for global-system-view stuff. With multiple non-global, composable drain hierarchies, logging does not have to just bunch of text messages scatted around the code that we print out and look at only when things go wrong.

Syslog works on lines of text, so it will have much harder job. `slog` has an actual structure of everything. In some cases it does not matter, in some syslog will not be able to do thing that slog can do easily. Applications that want to use syslog can still log to it using slog-syslog. 

Libraries most probably should not define their own drains (since it's up to application to decide where and how to log things), so application can choose to use async/sync io as it see fits.

Also, I have an idea for a `Streamer` (converter of strucutured loggiing information into binary/textual representation) that would print out stuff without "repeating" unnecessary context again, until it's actually different then the previous one.

**Comment by neithernut** [comment]
@dpc:
yes, syslog entries are primarily lines of text (plus a bit of meta information). And yes, bare syslog doesn't give you any tools for supplying contextual and structural information. If you want something like this, you have to build it yourself on top of syslog.

However, I still disregard the idea of manually logging to yet another file in a custom format. Consider the following use cases: in a fast-lived process with which the user interacts directly, you can happily log to stderr. The user will, in most cases, detect errors directly (because the programm tells the user) and react accordingly. You don't want some file lying around, which grows with time, that isn't ever looked at.
With long-running services however, you usually want some logging (e.g. to reconstruct the cause of errorneous behaviour). However, usually system administrators do not want a bunch of files in arbitrary locations and in arbitrary formats, but one, searchable repository of messages (except maybe Windows admins, but they are used the pain/do not know better). An entire set of tooling already exists for handling those logs.
Syslog daemons can log to a distant location or to a database. Syslog/logrotate can move/compress logs for archival, remove them after a certain amount of time or send an email to an administrator if something throws a critical error. Services exist for analyzing and evaluating those logs. There are even automization solutions based on syslog, easing the buren on system administrators.
Introducing each of those features to every logging librara there is would simply be insane.

That being said, it seems like slog can at least be used in a sane way, though I didn't have a closer look. As for avoiding repetition in the logs: I'd consider this a key feature.
